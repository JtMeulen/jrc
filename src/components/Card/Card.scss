@use '../../style/lib.scss' as *;
@use '../../style/mixins.scss';

.card {
  @include mixins.font-family();
}
