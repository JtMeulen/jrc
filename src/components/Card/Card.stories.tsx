import React from "react";
import { Meta, StoryFn } from "@storybook/react";
import Card from "./Card";

export default {
  title: "Components/Card",
  component: Card
} as Meta<typeof Card>;

const Template: StoryFn<typeof Card> = (args) => (
  <Card {...args} />
);

export const Primary = Template.bind({});
Primary.args = {};

