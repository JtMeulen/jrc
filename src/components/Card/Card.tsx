import React, { KeyboardEvent, ReactNode } from "react";
import cn from "classnames";
import "./Card.scss";

interface CardProps {
  children: ReactNode;
  headerImage?: {
    src: string;
    alt: string;
  };
  actionable?: boolean;
  onClick?: () => void;
}

const Card: React.FC<CardProps> = ({
  children,
  headerImage,
  actionable = false,
  onClick = () => {},
}) => {
  const handleonKeyDown = (event: KeyboardEvent) => {
    if (event.key === "Enter") {
      onClick();
    }
  };

  return (
    <article
      className={cn("card", { "card--actionable": actionable })}
      onClick={onClick} // TODO: Should we prevent onClick to be called if not actionable?
      onKeyDown={handleonKeyDown}
      tabIndex={actionable ? 0 : -1}
      data-testid="jrc-card"
    >
      {headerImage && (
        <img
          src={headerImage.src}
          alt={headerImage.alt}
          className="card__header-image"
        />
      )}
      <section>{children}</section>
    </article>
  );
};

export default Card;
