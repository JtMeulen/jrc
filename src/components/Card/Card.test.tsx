import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Card from "./Card";

describe("Card", () => {
  test("renders without error", () => {
    const { container } = render(
      <Card>
        <div></div>
      </Card>
    );
    expect(container).toMatchSnapshot();
  });

  test("renders with a header image", async () => {
    const { container } = render(
      <Card headerImage={{ src: "http://image.com", alt: "alt text" }}>
        <div></div>
      </Card>
    );
    expect(container).toMatchSnapshot();
  });

  test("is focusable when actionable", async () => {
    const { findByTestId } = render(
      <Card actionable>
        <div></div>
      </Card>
    );

    const cardEl = await findByTestId("jrc-card");

    await userEvent.keyboard("[Tab]");

    expect(cardEl).toHaveFocus();
  });

  test("calls onClick handler when clicked", async () => {
    const handleOnClick = jest.fn();
    const { findByTestId } = render(
      <Card actionable onClick={handleOnClick}>
        <div></div>
      </Card>
    );

    const cardEl = await findByTestId("jrc-card");

    fireEvent.click(cardEl);
    expect(handleOnClick).toHaveBeenCalled();
  });

  test("calls onClick handler when actionable and Enter key is pressed", async () => {
    const handleOnClick = jest.fn();
    render(
      <Card actionable onClick={handleOnClick}>
        <div></div>
      </Card>
    );

    await userEvent.keyboard("[Tab]");
    await userEvent.keyboard("[Enter]");
    expect(handleOnClick).toHaveBeenCalled();
  });
});
