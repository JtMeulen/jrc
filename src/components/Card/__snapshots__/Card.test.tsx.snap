// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders with a header image 1`] = `
<div>
  <article
    class="card"
    data-testid="jrc-card"
    tabindex="-1"
  >
    <img
      alt="alt text"
      class="card__header-image"
      src="http://image.com"
    />
    <section>
      <div />
    </section>
  </article>
</div>
`;

exports[`Card renders without error 1`] = `
<div>
  <article
    class="card"
    data-testid="jrc-card"
    tabindex="-1"
  >
    <section>
      <div />
    </section>
  </article>
</div>
`;
